### ISSUE TITLE
Roslyn does not take anonymous types' properties' nullability into account when inferring type arguments

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#35275]
**Version Used**: 

Branch master (25 Apr 2019)
Latest commit 95f137 by Gen Lu:
Merge pull request #35245 from genlu/FixMockExpService

Fix mock experiment service

**Steps to Reproduce**:

```cs
#nullable enable
class C
{
  void M1()
  {
    var x = new { A = (string?) null };

    if (x.A != null)
    {
      x.A.ToString(); // no warnings
      var y = new { x };
      y.x.A.ToString(); // no warnings
      
      var z = Id(x);
      z.A.ToString(); // warning
    }
  }
    
  T Id<T>(T t) => t;
}
```

**Expected Behavior**:
No warnings.

**Actual Behavior**:
`warning CS8602: Dereference of a possibly null reference.` is reported for `z.A.ToString()`;

**Notes**
The compiler knows that variable `x` contains an anontmous type `{A:string!}` with non-nullable property after the null check and takes this into account e.g. when inferring the result type of `new { x }` expression. However it does not take it into account when `x` is used in type inference.

**Comment by cston** [comment]
@TessenR, thanks for reporting this issue.

The current behavior is by design. The type of `x` is an anonymous type with property `string? A`. Within the `if` block, the value of `x.A` is not nullable. The type of `y.x.A` and `z.A` is `string?`, from the type of `x.A`, and the inferred nullability of `y.x.A` is inherited from `x.A` from the assignment, but the nullability of `z.A` cannot be inferred.
